### PR TITLE
Features/input components

### PR DESCRIPTION
**Pull request by VinayKashyap15**
Input components Added:
Custom input component : Scriptable object
Keyboard Component
InputManager SINGLETON class
implemented manager actions by using refernce

### CODE REVIEW

File: Assets/Scripts/Player/PlayerService.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace Player
+{
+    public class PlayerService : SingletonBase<PlayerService>
+    {
+
+        
+        public List<PlayerController> listOfPlayerControllers = new List<PlayerController>();
+        [SerializeField]
+        private InputScriptableObjectList listOfInputs;
```
Comment (mayankgrover) [inline Assets/Scripts/Player/PlayerService.cs]: the type of input for each player should be part of the Player Scriptable object

File: Assets/Scripts/Player/PlayerView.cs [hunk 2]
```diff
@@ -8,30 +10,41 @@ public class PlayerView : MonoBehaviour
 
         [SerializeField]
         private GameObject muzzlePoint;
-
+        private float currentScore;
```
Comment (mayankgrover) [inline Assets/Scripts/Player/PlayerView.cs]: current score is data - should be maintained in the Model not View

File: Assets/Scripts/Weapons/Bullet/BulletService.cs [hunk 3]
```diff
@@ -8,15 +9,20 @@ namespace Weapons.Bullet
...
         public BULLET_TYPE typeOfBullet;
-
+        private PlayerController playerControllerInstance;
         public float GetBulletSpeed(BulletModel _model)
         {
             return _model.GetBulletSpeed();
         }
 
-        public BulletController SpawnBullet()
+        public PlayerController GetPlayerControllerInstance()
```
Comment (mayankgrover) [inline Assets/Scripts/Weapons/Bullet/BulletService.cs]: this shouldn't be here. No one should be asking BulletService to get player controller. If someone wants player controller they should go to PlayerService

File: Assets/Scripts/Weapons/Bullet/ControllerScripts/BulletController.cs [hunk 4]
```diff
@@ -1,6 +1,9 @@
 ﻿using UnityEngine;
+using Common;
+using Player;
 using Bullet.ModelScripts;
 using Bullet.ViewScripts;
+using Weapons.Bullet;
 
 namespace Bullet.ControllerScripts
```
Comment (mayankgrover) [inline Assets/Scripts/Weapons/Bullet/ControllerScripts/BulletController.cs]: can be renamed to just Bullet.Controller (in general no need to use  "Scripts" as a suffix or prefix for namespaces or cs file names etc)

File: Assets/Scripts/Weapons/Bullet/ControllerScripts/BulletController.cs [hunk 5]
```diff
@@ -43,5 +48,22 @@ public virtual void StartDestroy()
         {
             currentBulletModel = null;
         }
+
+        public void FireBullet(Vector3 _firePosition, Quaternion _fireRotation, Vector3 _fireDirection)
+        {
+            bulletInstance.transform.position = _firePosition;
```
Comment (mayankgrover) [inline Assets/Scripts/Weapons/Bullet/ControllerScripts/BulletController.cs]: setting transform position/velocity/rotation/etc anything to do with GO should be inside view

File: Assets/Scripts/Player/PlayerService.cs (hunk 1 above)
Comment (VinayKashyap15) [reply Assets/Scripts/Player/PlayerService.cs]: Do you mean expose the type of input component as an enum?
or create three different types of scriptable objects?

File: Assets/Scripts/Player/PlayerView.cs (hunk 2 above)
Comment (VinayKashyap15) [reply Assets/Scripts/Player/PlayerView.cs]: okay done

File: Assets/Scripts/Weapons/Bullet/ControllerScripts/BulletController.cs (hunk 4 above)
Comment (VinayKashyap15) [reply Assets/Scripts/Weapons/Bullet/ControllerScripts/BulletController.cs]: Okay

File: Assets/Scripts/Common/InputManagerBase.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿using InputComponents;
+using Player;
```
[the remark this answers is not in the document]
Comment (VinayKashyap15) [reply Assets/Scripts/Common/InputManagerBase.cs]: I removed it from commons folder and added it to input components folder.

File: Assets/Scripts/Common/ScoreManager.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using UnityEngine;
+using Player;
```
[the remark this answers is not in the document]
Comment (VinayKashyap15) [reply Assets/Scripts/Common/ScoreManager.cs]: I added this class for all UI stuff across all levels. Currently I'm making this class empty. I have removed the unnecessary using statements.

File: Assets/Scripts/Player/PlayerService.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        private GameObject playerPrefab;
+        private GameObject playerInstance;
+        int playerID = 0;
+
+        private GameObject SpawnPrefabInstance(Vector3 _spawnPos)
+        {
+            if (!playerPrefab)
+            {
+                playerPrefab = Resources.Load("Player") as GameObject;
```
[the remark this answers is not in the document]
Comment (VinayKashyap15) [reply Assets/Scripts/Player/PlayerService.cs]: I can do that as well.

File: Assets/Scripts/Player/PlayerService.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+        private void SpawnPlayers()
+        {
+            PlayerController _playerControllerInstance;
+            if (listOfInputs)
+            {
+                Vector3 pos = new Vector3(0, 0, 0);
+               
+                
+                for (int i = 0; i < listOfInputs.inputList.Count; i++)
```
[the remark this answers is not in the document]
Comment (VinayKashyap15) [reply Assets/Scripts/Player/PlayerService.cs]: Currently I'm treating a new input component as a new player itself. I understand what you pointed out and will try to rectify this.

File: Assets/Scripts/Weapons/Bullet/BulletService.cs (hunk 3 above)
Comment (VinayKashyap15) [reply Assets/Scripts/Weapons/Bullet/BulletService.cs]: done

File: Assets/Scripts/Player/PlayerController.cs [hunk 10]
```diff
@@ -1,41 +1,64 @@
...
+            playerModel.SetID(_playerID);
             playerView = playerViewInstance;
+            
+            
+            currentInputComponent = new KeyboardComponent(this);
+
         }
-        public void Move(float h, float v)
+
+        public PlayerController(PlayerView playerViewInstance, InputScriptableObject _customInputScheme,int _playerID)
```
[the remark this answers is not in the document]
Comment (VinayKashyap15) [reply Assets/Scripts/Player/PlayerController.cs]: Done

File: Assets/Scripts/Player/PlayerService.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (VinayKashyap15) [reply Assets/Scripts/Player/PlayerService.cs]: Done

**Pull request merged**